### PR TITLE
Rebase to v2.36.0-rc0

### PR DESCRIPTION
**Pull request by vdye**
## Notable changes
- Lots of conflicts in `object-file.c` between upstream `ab/object-file-api-updates` and local branch `PhilipOakley/hashliteral_t`. I *think* I caught all the places that needed to change to `size_t`, but I'd appreciate a second opinion.
- Per fixups on `shears/main`:
  - Dropped `ns/batched-fsync` branch (see f070b3f1d4)
  - Applied 351453516d & d06c659461
- Re-applied FSMonitor per `@jeffhostetler`'s feedback
  - Removed all of the local branches `try-v4-fsmonitor-part2`, `try-v4-fsmonitor-part3`, `try-v4-fsmonitor-part4`, `jeffhostetler-try-v4-fsmonitor-part5`, and `mark-v4-fsmonitor-experimental`
  - Replaced the above with the [rebased version](https://github.com/jeffhostetler/git/tree/trial-fsmonitor-march-2022-part3) of the upstream-submitted V3
  - Replaced e50ce8c06b with cc9c45eb2a5c17197cd988bd27845da9392f5bc0
  - Cleaned up some usage strings in `t/helper/test-fsmonitor-client.c` 
- Other misc formatting conflicts (i.e., moving around code without functional changes)

## Range-diff
See ~~[here](https://github.com/git-for-windows/git/files/8412812/pr-range-diff.txt) [here](https://github.com/git-for-windows/git/files/8413309/pr-range-diff.txt)~~ [HERE (hopefully the last one)](https://github.com/git-for-windows/git/files/8419491/pr-range-diff.txt) - it's too long for the PR description 🙁

### CODE REVIEW

**Comment by vdye** [comment]
@dscho Found the Windows build issue (I didn't apply one of the fixups correctly - it should have removed a201fadee40be99154dbd3b06241c2ff897ac242, but I misinterpreted the results). I've pushed the update and updated the range-diff in the PR description.

**Comment by dscho** [comment]
Wow, this was super fast!

> * Lots of conflicts in `object-file.c` between upstream `ab/object-file-api-updates` and local branch `PhilipOakley/hashliteral_t`. I _think_ I caught all the places that needed to change to `size_t`, but I'd appreciate a second opinion.

Looks good to me! I resolved the same conflicts in `shears/seen` a few weeks ago and your version looks pretty similar to mine. I have not verified this in detail, but as long as it compiles under `DEVELOPER=1` without any error, I am confident that it is correct.

> * Dropped `ns/batched-fsync` branch (see [f070b3f](https://github.com/git-for-windows/git/commit/f070b3f1d4436836216dd459a287dd77f1671bf2))

Could we maybe merge `ns/batch-fsync` instead, i.e. what was merged in 866ee261d43d? That would address this part of the range-diff:

> ```
> -252:  a3ba62730e <   -:  ---------- bulk-checkin: rename 'state' variable and separate 'plugged' boolean
> [...]
> -255:  b15e2b781d <   -:  ---------- core.fsyncobjectfiles: batched disk flushes
> -256:  6be6d616cc <   -:  ---------- core.fsyncobjectfiles: add windows support for batch mode
> -257:  55442b7e77 <   -:  ---------- update-index: use the bulk-checkin infrastructure
> -258:  e7fd484006 <   -:  ---------- unpack-objects: use the bulk-checkin infrastructure
> -259:  6b15246b48 <   -:  ---------- core.fsyncobjectfiles: tests for batch mode
> -260:  7bc1fe5538 <   -:  ---------- core.fsyncobjectfiles: performance tests for add and stash
> ```

About this:

> See [here](https://github.com/git-for-windows/git/files/8412704/pr-range-diff.txt) - it's too long for the PR description 🙁

Right, it is dominated by the `size_t` and the FSMonitor updates. Speaking of which:

> ```
> - 87:  6d867a37b2 <   -:  ---------- fsmonitor: update fsmonitor config documentation
> [...]
> - 91:  c9b110ec39 <   -:  ---------- builtin/fsmonitor--daemon: use parse-options API fully
> [...]
> -115:  ec8e0b6389 <   -:  ---------- t/perf/p7519: add fsmonitor--daemon test cases
> [...]
> -132:  e5dfb96e11 <   -:  ---------- fsmonitor--daemon: print start message only if fsmonitor.announceStartup
> -142:  cd80dd2917 <   -:  ---------- fsmonitor: measure time taken to apply fsmonitor query result
> ```

These were actually squashed into other commits.

> -148:  54427b1cde <   -:  ---------- fsmonitor: mark the built-in FSMonitor as experimental

Yaaaay!

However, I would like to see (now-deprecated) support for `core.useBuiltinFSMonitor` come back, and if it is set to `true` without `core.fsmonitor` being set to `true` also, I would like to see some advice being shown.

Related, dfbbfc7e5c37967c050a741dd7787abd92091d42 seems to rely on `core.useBuiltinFSMonitor`, still, and should probably be updated to use `core.fsmonitor` instead.

> ```
> -265:  1383ae5f12 <   -:  ---------- gpg-interface: trim CR from ssh-keygen -Y find-principals
> ```

Right, this was done differently upstream, but the original bug is still fixed.

**Comment by dscho** [comment]
@vdye the force-push looks good. Please note that my previous comment still talks about the pre-push range-diff ;-)

**Comment by dscho** [comment]
> Could we maybe merge `ns/batch-fsync` instead, i.e. what was merged in [866ee26](https://github.com/git-for-windows/git/commit/866ee261d43df5ecfc12ac24af9399dc2ba13b72)?

I just noticed that I needed 9f8eb4a9cf8f56d841d1f18c7e385236223dfd4a after cherry-picking 5458f300f6e6 into `shears/next` to make it compile again... I'm not 100% positive that this is the correct fix, though.

**Comment by vdye** [comment]
> However, I would like to see (now-deprecated) support for `core.useBuiltinFSMonitor` come back, and if it is set to `true` without `core.fsmonitor` being set to `true` also, I would like to see some advice being shown.
> 

@dscho Do you already have an implementation for this on one of the `shears/` branches? Or should I make this change myself?

Otherwise, everything else you suggested makes sense - once I get things to build, I'll push & update the range-diff.

**Comment by vdye** [comment]
> I just noticed that I needed [9f8eb4a](https://github.com/git-for-windows/git/commit/9f8eb4a9cf8f56d841d1f18c7e385236223dfd4a) after cherry-picking [5458f30](https://github.com/git-for-windows/git/commit/5458f300f6e642557a1ffbdd2e66b967edba326c) into `shears/next` to make it compile again... I'm not 100% positive that this is the correct fix, though.

This might be specific to `next`, since I'm [not seeing](https://github.com/vdye/git/actions/runs/2092484451) the same failures when I cherry-pick `ns/batch-fsync`.

Since it _appears_ to be working without the extra fixup, I pushed the changes & updated the range-diff. Re-introducing `core.useBuiltinFSMonitor` is still pending since 1) we should decide if it's needed (and with what level of deprecation warning, if any), and 2) I need to implement it.

**Comment by vdye** [comment]
@dscho One more update! I added a commit (9b32887c37) to re-introduce `core.useBuiltinFSMonitor` (including a deprecation warning), and updated b85b04f3b3, which now requires that neither `core.useBuiltinFSMonitor` _and_ `core.fsmonitor` are unset to define the "default" behavior. 

The build is passing & all of your earlier comments are addressed, so I'm going to trigger the artifacts build. If something else needs to change, that's fine, but on the off chance that this looks okay, we won't have to wait for the artifacts to build to continue with the release :)

**Comment by dscho** [comment]
> > However, I would like to see (now-deprecated) support for `core.useBuiltinFSMonitor` come back, and if it is set to `true` without `core.fsmonitor` being set to `true` also, I would like to see some advice being shown.
> 
> @dscho Do you already have an implementation for this on one of the `shears/` branches?

Sadly, I did not...

> > I just noticed that I needed [9f8eb4a](https://github.com/git-for-windows/git/commit/9f8eb4a9cf8f56d841d1f18c7e385236223dfd4a) after cherry-picking [5458f30](https://github.com/git-for-windows/git/commit/5458f300f6e642557a1ffbdd2e66b967edba326c) into `shears/next` to make it compile again... I'm not 100% positive that this is the correct fix, though.
> 
> This might be specific to `next`, since I'm [not seeing](https://github.com/vdye/git/actions/runs/2092484451) the same failures when I cherry-pick `ns/batch-fsync`.

My mistake. I only cherry-picked a single commit, but left the rest unchanged (because they did not conflict). What you did is correct, what I did isn't.

File: fsmonitor-settings.c
Comment (dscho) [inline fsmonitor-settings.c:107]: Just like `core.fsmonitor` overrides `GIT_TEST_FSMONITOR`, we now also let `core.useBuiltinFSMonitor` override it. Good!

File: repo-settings.c
Comment (dscho) [inline repo-settings.c:62]: If I read this correctly, FSMonitor is now enabled upon `experimental.manyFile` unless _both_ `core.fsmonitor` _and_ `core.useBuiltinFSMonitor` are configured (and it overrides `core.fsmonitor=<path>`). I think it should read like this instead:

```c
		if (manyfiles &&
		    fsmonitor_ipc__is_supported()  &&
		    fsm_settings__get_mode(r) == FSMONITOR_MODE_DISABLED &&
		    repo_config_get_maybe_bool(r, "core.fsmonitor", &value) ==1 /* unset */ &&
		    repo_config_get_bool(r, "core.useBuiltinFSMonitor", &value) /* also unset */)
```

Comment (vdye) [reply repo-settings.c]: Fixed, now in commit 3dd2f446b5

**Comment by vdye** [comment]
> @vdye your "hopefully the last one" range-diff file appears to be out-of-date. I couldn't get the commit for this line:
> 
> ```
>   -:  ---------- > 158:  9b32887c37 fsmonitor: reintroduce core.useBuiltinFSMonitor
> ```
> 
> (and the remaining lines are not reachable from my fetch, either.)

I had accidentally re-attached an old version of the file 🙁 New one is now attached, and the commit that was 9b32887c37 is now a28a3c3326.

**Comment by jeffhostetler** [comment]
I think this looks good.  It passed my FSMonitor tests on both Windows and Mac.

**Review by derrickstolee — APPROVED** [review]
I took a look at the range-diff and it looks good.  There's a lot in there, due to FS Monitor, but I don't think we are going to catch anything else during the pre-`rc0` review.

Let's get this released and get started on the `microsoft/git` rebase so we can get some additional tests on this.

**Review by dscho — APPROVED** [review]
Very good! Let's :shipit:

**Pull request merged**
